raftAttributes key doesn't exist

I have a 3 node cluster set up for testing and was attempting some load tests yesterday (thousands of writes per second with key watchers going too), when I got an error I haven't seen before:

```
2015/02/13 08:18:11 etcd: listening for peers on http://10.126.75.72:7001
2015/02/13 08:18:11 etcd: listening for peers on http://127.0.0.1:7001
2015/02/13 08:18:11 etcd: listening for client requests on http://10.126.75.72:4001
2015/02/13 08:18:11 etcd: listening for client requests on http://127.0.0.1:4001
2015/02/13 08:18:11 etcdserver: discovery token ignored since a cluster has already been initialized. Valid log found at "/var/lib/etcd/mvp01.etcd/wal"
2015/02/13 08:18:12 etcdserver: recovered store from snapshot at index 172422
2015/02/13 08:18:12 nodeToMember should never fail: raftAttributes key doesn't exist
panic: nodeToMember should never fail: raftAttributes key doesn't exist

goroutine 1 [running]:
log.Panicf(0x8b0090, 0x22, 0xc20a049040, 0x1, 0x1)
        /usr/local/go/src/log/log.go:314 +0xd0
github.com/coreos/etcd/etcdserver.membersFromStore(0x7f4f9f68a578, 0xc2080621e0, 0xc2080621e0, 0xc208cae140)
        /Users/philips/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/etcdserver/cluster.go:362 +0x398
github.com/coreos/etcd/etcdserver.NewClusterFromStore(0x7fff6667578d, 0x3a, 0x7f4f9f68a578, 0xc2080621e0, 0x1)
        /Users/philips/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/etcdserver/cluster.go:101 +0x179
github.com/coreos/etcd/etcdserver.NewServer(0xc20803a210, 0xc208058480, 0x0, 0x0)
        /Users/philips/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server.go:213 +0x20b6
github.com/coreos/etcd/etcdmain.startEtcd(0xc208082000, 0x0, 0x0, 0x0)
        /Users/philips/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/etcdmain/etcd.go:161 +0x18cc
github.com/coreos/etcd/etcdmain.Main()
        /Users/philips/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/etcdmain/etcd.go:54 +0x1fc
main.main()
        /Users/philips/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/main.go:31 +0x1b
```

Possibly pertinent info:
 - etcd 2.0
 - using public discovery service
 - all 3 nodes error like this when restarted
 - one node is running in a docker container


EDIT: Wrapping ugliness

UPDATE:  I was able to get the cluster back up by deleting the data-dirs on all nodes.

--- a/etcdserver/cluster.go
+++ b/etcdserver/cluster.go
@@ -346,6 +346,20 @@ func (c *Cluster) UpdateRaftAttributes(id types.ID, raftAttr RaftAttributes) {
 	c.members[id].RaftAttributes = raftAttr
 }
 
+// Validate ensures that there is no identical urls in the cluster peer list
+func (c *Cluster) Validate() error {
+	urlMap := make(map[string]bool)
+	for _, m := range c.Members() {
+		for _, url := range m.PeerURLs {
+			if urlMap[url] {
+				return fmt.Errorf("duplicate url %v in cluster config", url)
+			}
+			urlMap[url] = true
+		}
+	}
+	return nil
+}
+
 func membersFromStore(st store.Store) (map[types.ID]*Member, map[types.ID]bool) {
 	members := make(map[types.ID]*Member)
 	removed := make(map[types.ID]bool)

--- a/etcdserver/config.go
+++ b/etcdserver/config.go
@@ -62,15 +62,8 @@ func (c *ServerConfig) VerifyBootstrapConfig() error {
 		return fmt.Errorf("initial cluster state unset and no wal or discovery URL found")
 	}
 
-	// No identical IPs in the cluster peer list
-	urlMap := make(map[string]bool)
-	for _, m := range c.Cluster.Members() {
-		for _, url := range m.PeerURLs {
-			if urlMap[url] {
-				return fmt.Errorf("duplicate url %v in cluster config", url)
-			}
-			urlMap[url] = true
-		}
+	if err := c.Cluster.Validate(); err != nil {
+		return err
 	}
 
 	// Advertised peer URLs must match those in the cluster peer list

--- a/etcdserver/server.go
+++ b/etcdserver/server.go
@@ -185,6 +185,9 @@ func NewServer(cfg *ServerConfig) (*EtcdServer, error) {
 			if cfg.Cluster, err = NewClusterFromString(cfg.Cluster.token, s); err != nil {
 				return nil, err
 			}
+			if cfg.Cluster.Validate() != nil {
+				return nil, fmt.Errorf("bad discovery cluster: %v", err)
+			}
 		}
 		cfg.Cluster.SetStore(st)
 		cfg.PrintWithInitial()

